contributors/design-proposals/mount-options.md

```diff
@@ -0,0 +1,105 @@
+
+Mount time options that are operationally important and have no security implications should be suppported. Examples are NFS's TCP mode, versions, lock mode, caching mode; Glusterfs's caching mode; SMB's version, locking, id mapping; and more.
+
+## Design
+
+Mount options can be specified as `mountOption`  as part of PV. As an example:
+
+``` yaml
+ apiVersion: v1
+  kind: PersistentVolume
+  metadata:
+    name: pv0003
+  spec:
+    capacity:
+      storage: 5Gi
+    accessModes:
+      - ReadWriteOnce
+    persistentVolumeReclaimPolicy: Recycle
+    nfs:
+      path: /tmp
+      server: 172.17.0.2
+      mountOption: "hard,nolock,nfsvers=3"
+```
+
+To be consistent with VolumeSource in pod specs, it can also be specified as part of Volume spec with Pod:
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test-pd
+spec:
+  containers:
+  - image: gcr.io/google_containers/test-webserver
+    name: test-container
+    volumeMounts:
+    - mountPath: /test-pd
+      name: test-volume
+  volumes:
+  - name: test-volume
+    nfs:
+      server: "192.168.1.100"
+      path: "/var/exports"
+      mountOption: "hard,nolock,nfsvers=3"
+```
+
+Or it can be specified as part of StorageClass definition:
+
+```yaml
+apiVersion: storage.k8s.io/v1beta1
+kind: StorageClass
+metadata:
+  name: slow
+provisioner: kubernetes.io/glusterfs
+parameters:
+  resturl: "http://127.0.0.1:8081"
+  restauthenabled: "true"
+  restuser: "admin"
+  secretNamespace: "default"
+  secretName: "heketi-secret"
+  mountOption: "volume-name=foobar"
+```
+
+## Preventing users from specifying invalid/unsupported mount options
+
+While mount options enable more flexibility in how volumes are mounted, it can result
+in user specifying options that are not supported or are known to be problematic.
+
+There could be several possibly alternatives for this. I am listing few I considered.
+
+
+1. Add support for blacklisting mount options as `kubelet` startup configuration:
+
+To prevent users from using options which are not safe - Kubernetes admin can configure volume options and specify a blacklist of mount point parameters. For example:
+
+```
+~> kubelet --volume-options <volume-config-file>
+```
+
+The volume configuration file may look like:
+
+```
+{
+    "glusterfs": {
+      "mount_option_blacklist":  ["transport"]
+    },
+    "nfs": {
+        "mount_option_blacklist": ["hard", "noacl"]
+    }
+}
+```
+
+To match options that are allowed - we will split the mount options in PV by ',' and then verify if key does not exist in blacklist, which will prevent mounting volumes with these options.
+
+One upside of using these options in kubelet is - it will also work for external provisioners. One major downside is - we will be allowing creation of PV or SC which can be blocked later by a blacklisted option. i.e - A PV or SC with invalid mount option will not be rejected immediately.
```

Comment:
Agreed on all accounts but so far use cases we have seen - there hasn't been a case where we have a need of some PVs from same plugin configurable with wider(?) mount options and some with narrower. 

The other options we have considered has similar problems. But this is still very much WIP - so we are considering all options.